### PR TITLE
127 implement review functionality

### LINKED ISSUE #127: Implement Review Functionality
Add review functionality for spring boot.

To Do:
- [x] Add Review Functionality
- [x] Delete Review Functionality

### PR DESCRIPTION
**Pull request by simphiwe-nonabe**
# Description
Adding A Review Functionality And Deleting A Review Functionality Implemented

Fixes #127 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: libs/app/recipe/feature/src/recipe.page.ts [hunk 1]
```diff
@@ -42,15 +48,25 @@ export class RecipePage implements OnInit {
...
-      (response: IRecipe) => {
-        this.recipe = response;
-      },
-      error => {
-        this.errorMessage = 'Error retrieving recipe data.';
-        this.recipe = undefined;
-      }
-    );
+    this.store.dispatch(new GetRecipe(id));
+    this.recipe$.pipe(take(1)).subscribe(stateRecipe => {
```
Comment (ryanbasiltrickett) [inline libs/app/recipe/feature/src/recipe.page.ts:52]: This is not needed

File: libs/app/recipe/data-access/src/recipe.api.ts [hunk 2]
```diff
@@ -12,7 +12,85 @@ export class RecipeService {
...
+        'Bread',
+        'Simple'
+      ]
+    }
+
     const url = `${this.baseUrl}/${id}`;
-    return this.http.get<IRecipe>(url);
+    //TODO: Uncomment on backend fix
+    //return this.http.get<IRecipe>(url);
+    return new BehaviorSubject<IRecipe>(dummyRecipe);
```
Comment (ryanbasiltrickett) [inline libs/app/recipe/data-access/src/recipe.api.ts:89]: This should be using the API and not mock data

File: libs/app/recipe/data-access/src/recipe.state.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    this.api.updateRecipe(recipe);
+  }
+
+  @Action(AddReview)
+  addRecipeReview({ getState } : StateContext<RecipeStateModel>, { review } : AddReview) {
+
+    let newRecipe = getState().recipe;
+
+    if(newRecipe){
+      newRecipe?.reviews?.unshift(review);
```
Comment (ryanbasiltrickett) [inline libs/app/recipe/data-access/src/recipe.state.ts:56]: You never patch the state

File: libs/app/recipe/feature/src/recipe.page.ts (hunk 1 above)
Comment (kphutiyagae) [reply libs/app/recipe/feature/src/recipe.page.ts:52]: The dispatch or the pipe?

File: libs/app/recipe/data-access/src/recipe.state.ts (hunk 3 above)
Comment (kphutiyagae) [reply libs/app/recipe/data-access/src/recipe.state.ts:56]: Unnecessary as it is done in the updateRecipe function. Patching state twice for the same action is not the best idea.

File: libs/app/recipe/data-access/src/recipe.state.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  }
+
+  @Action(DeleteReview)
+  removeRecipeReview({ getState } : StateContext<RecipeStateModel>, { reviewId } : DeleteReview) {
+
+    let newRecipe = getState().recipe;
+
+    if(newRecipe){
+      newRecipe.reviews = newRecipe?.reviews?.filter( currentReview => currentReview.reviewId !== reviewId);
```
[the remark this answers is not in the document]
Comment (kphutiyagae) [reply libs/app/recipe/data-access/src/recipe.state.ts:69]: Please see previous comment.

File: libs/app/recipe/feature/src/recipe.page.ts (hunk 1 above)
Comment (ryanbasiltrickett) [reply libs/app/recipe/feature/src/recipe.page.ts:52]: The subscription cause you have a selector

Comment (kphutiyagae) [reply libs/app/recipe/feature/src/recipe.page.ts:52]: But it is an Observable, and Observables only emit upon subscription.

File: libs/app/recipe/data-access/src/recipe.api.ts (hunk 2 above)
Comment (kphutiyagae) [reply libs/app/recipe/data-access/src/recipe.api.ts]: Done.

**Review by ryanbasiltrickett — APPROVED** [review]

**Pull request merged**